postfix_mta_sts_resolver/proactive_fetcher.py

```diff
@@ -0,0 +1,113 @@
+                if status is STSFetchResult.VALID:
+                    pol_id, pol_body = policy
+                    cached = CacheEntry(time.time(), pol_id, pol_body)
+                    await self._cache.safe_set(domain, cached, self._logger)
+                elif status is STSFetchResult.FETCH_ERROR or status is STSFetchResult.NONE:
+                    self._logger.warning("Domain %s does not have a valid policy.", domain)
+            except asyncio.CancelledError:  # pragma: no cover pylint: disable=try-except-raise
+                raise
+            except Exception as exc:  # pragma: no cover
+                self._logger.exception("Unhandled exception: %s", exc)
+            finally:
+                domain_queue.task_done()
+
+    async def iterate_domains(self):
+        self._logger.info("Proactive policy fetching "
+                          "for all domains in cache started...")
+
+        # Create domain processor tasks
+        domain_processors = []
+        domain_queue = asyncio.Queue(maxsize=constants.DOMAIN_QUEUE_LIMIT)
+        for _ in range(constants.DOMAIN_REQUEST_LIMIT):
+            domain_processor = self._loop.create_task(self.process_domain(domain_queue))
+            domain_processors.append(domain_processor)
+
+        # Produce work for domain processors
+        try:
+            token = None
+            while True:
+                token, cache_items = await self._cache.scan(token, constants.DOMAIN_QUEUE_LIMIT)
+                self._logger.debug("Enqueued %d domains for processing.", len(cache_items))
+                for cache_item in cache_items:
+                    await domain_queue.put(cache_item)
+                if token is None:
+                    break
+
+            # Wait for queue to clear
+            await domain_queue.join()
+        # Clean up the domain processors
+        finally:
+            for domain_processor in domain_processors:
+                domain_processor.cancel()
+            await asyncio.gather(*domain_processors, return_exceptions=True)
+
+        # Update the proactive fetch timestamp
+        await self._cache.set_proactive_fetch_ts(time.time())
+
+        self._logger.info("Proactive policy fetching "
+                          "for all domains in cache finished.")
+
+    async def fetch_periodically(self):
+        async def fetch():
+            domain_iteration_future = self._loop.create_task(self.iterate_domains())
+
+            try:
+                await asyncio.shield(domain_iteration_future)
+            except asyncio.CancelledError:  # pragma: no cover
+                try:
+                    await asyncio.wait_for(domain_iteration_future, self._shutdown_timeout)
+                except asyncio.TimeoutError:  # pragma: no cover
+                    self._logger.warning("Shutdown timeout expired. "
+                                         "Shutting down forcefully.")
+                    await domain_iteration_future
```

Comment:
Hmm, my intended behaviour was this:
The shield indeed protects `domain_iteration_future` from cancellation, but only until the `wait_for` expires, which then forwards the cancellation to the future/task. I put the `await domain_iteration_future` there to normalise the behaviour that changed in py3.7: https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for (as I understand it, this specific `await domain_iteration_future` should be a no-op in py3.7 and later, due to `wait_for` already waiting for cancellation).

So my goal here was to give the proactive fetcher a chance to finish its work before being killed off, essentially trying to be as graceful as possible in the shutdown. It wasn't meant to wait indefinitely, but only for the allowed graceful (`self._shutdown_timeout`). Did I miss something?

Reply:
Ah, now I get it. I forgot `wait_for` cancels task on timeout and confused it with `asyncio.wait` :+1: 

As it appears to me, these few seconds of grace for shutdown make difference when duration of prefetch run is comparable to shutdown timeout itself (i.e. it lasts few seconds and small delay has high chance to let it finish). In other cases it just slows down shutdown or restart of daemon, leaving policy server unresponsive for a longer period. Graceful shutdown makes sense for responder component because if postfix is querying domain policy right now it's better to let it answer, otherwise it'll apply no TLS policy outgoing connection. It's not the case for cache prefetch: if it gets terminated it's not a big deal, it'll start iterating domains again right after restart.

Reply:
Yup, fair point, uptime is more important than saving resources usage here.

Reply:
@galme However, you still can save resources here. In this case of unfinished cache prefetch new run will re-fetch again entries which were already fetched in last interrupted run. You can avoid it if you'll skip cache entries which age is less than half of prefetching interval. It seems reasonable and new cache fetch will practically start almost from same point of interruption.